### ISSUE TITLE
Dev: allow tests to pass, even when not all tool dependencies are installed

### ISSUE THREAD

**Issue by twpayne** [goreleaser/goreleaser#5104]
### Is your feature request related to a problem? Please describe.

Running

```console
$ git clone https://github.com/goreleaser/goreleaser.git
...
$ cd goreleaser
$ go test ./...
```

on a newly-installed machine fails due to multiple missing binaries and services, e.g.:

```
2024/08/27 19:12:44 Get "http://unix.sock/_ping": dial unix /var/run/docker.sock: connect: no such file or directory
FAIL	github.com/goreleaser/goreleaser/v2/internal/pipe/blob	0.020s
...
  • nix-prefetch-url is not available
  • nix-prefetch-url is not available
--- FAIL: TestSkip (0.00s)
    --- FAIL: TestSkip/nix-all-good (0.00s)
        nix_test.go:38: 
            	Error Trace:	/home/twp/src/github.com/goreleaser/goreleaser/internal/pipe/nix/nix_test.go:38
            	Error:      	Should be false
            	Test:       	TestSkip/nix-all-good
  • fake-nix-prefetch-url is not available
  • nix-prefetch-url is not available
--- FAIL: TestPrefetcher (0.00s)
    --- FAIL: TestPrefetcher/prefetch (0.00s)
        --- FAIL: TestPrefetcher/prefetch/publish (0.00s)
            --- FAIL: TestPrefetcher/prefetch/publish/valid (0.00s)
                nix_test.go:66: 
                    	Error Trace:	/home/twp/src/github.com/goreleaser/goreleaser/internal/pipe/nix/nix_test.go:66
                    	Error:      	Received unexpected error:
                    	            	could not prefetch url: https://github.com/goreleaser/goreleaser/releases/download/v1.18.2/goreleaser_Darwin_arm64.tar.gz: exec: "nix-prefetch-url": executable file not found in $PATH: 
                    	Test:       	TestPrefetcher/prefetch/publish/valid
    --- FAIL: TestPrefetcher/available (0.00s)
        --- FAIL: TestPrefetcher/available/publish (0.00s)
            --- FAIL: TestPrefetcher/available/publish/valid (0.00s)
                nix_test.go:80: 
                    	Error Trace:	/home/twp/src/github.com/goreleaser/goreleaser/internal/pipe/nix/nix_test.go:80
                    	Error:      	Should be true
                    	Test:       	TestPrefetcher/available/publish/valid
...
--- FAIL: TestRun (0.91s)
    upx_test.go:119: 
        	Error Trace:	/home/twp/src/github.com/goreleaser/goreleaser/internal/pipe/upx/upx_test.go:119
        	Error:      	Received unexpected error:
        	            	upx not found in PATH
        	Test:       	TestRun
FAIL
FAIL	github.com/goreleaser/goreleaser/v2/internal/pipe/upx	1.728s
```

goreleaser's dev requirements are clearly documented in

https://github.com/goreleaser/goreleaser/blob/a9e0a8f11209b992b62e126274fb958e4527d65f/CONTRIBUTING.md?plain=1#L15-L23

However, requiring all of these raises the barrier to entry for contributors, and irrelevant failing tests add noise and time cost (they are re-run) for contributors who know which failures can be ignored.

### Describe the solution you'd like

Tests should be skipped if the required dependencies are not installed, for example:

```go
if _, err := os.LookPath("upx"); err != nil {
    t.Skip("upx not found in path")
}
```

Goreleaser's extensive CI will still catch problems, even if the PR author skips tests.

### Describe alternatives you've considered

Installing all the dependencies (painful: and some require non-trivial changes to the development machine, e.g. Docker).

Ignoring test failures (painful: the tests that fail due to my change are hard to spot amongst multiple other unrelated test failures).

Running `go test` just on the affected packages (painful: I need to work out which tests are relevant, and might get this wrong).

### Search

- [X] I did search for other open and closed issues before opening this

### Supporter

- [ ] I am a [sponsor](https://github.com/sponsors/caarlos0/) or a [Pro](https://goreleaser.com/pro) customer

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Additional context

I am happy to submit a PR for this. Just say!

**Comment by caarlos0** [comment]
good call! I think I did that for some things, but not all of them... will do!

**Comment by twpayne** [comment]
Thank you for the fast fix!

There's still a dependency on docker which is not addressed by https://github.com/goreleaser/goreleaser/commit/599ce44c749ef13b4e6a3042ece90aa3e3f3dcaa. Maybe docker is a sufficiently fundamental dependency that it is actually a requirement.

**Comment by caarlos0** [comment]
> Thank you for the fast fix!
> 
> There's still a dependency on docker which is not addressed by [599ce44](https://github.com/goreleaser/goreleaser/commit/599ce44c749ef13b4e6a3042ece90aa3e3f3dcaa). Maybe docker is a sufficiently fundamental dependency that it is actually a requirement.

oh, let me take a look

**Comment by caarlos0** [comment]
ah, good catch, https://github.com/goreleaser/goreleaser/pull/5109
